File: dart/test/codecs_test.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:multipath/multipath.dart';
import 'package:test/test.dart';

List<Frame> drainByteByByte(Uint8List bytes) {
  final r = FrameReader();
  final frames = <Frame>[];
  for (final b in bytes) {
    r.push(Uint8List.fromList([b]));
    Frame? f;
    while ((f = r.next()) != null) {
      frames.add(f!);
    }
  }
  return frames;
}

void main() {
  test('redundant frames round-trip, reassembled byte by byte', () {
    final connId = Uint8List.fromList(List.generate(16, (i) => i + 1));
    final payload = Uint8List.fromList(List.generate(300, (i) => i & 0xff));
    final b = BytesBuilder();
    b.add(encodeHello(connId, 7));
    b.add(encodeData(1234567890123, payload));
    b.add(encodeAck(999));
    b.add(encodeNonce(framePing, 42));
    b.add(encodeNonce(framePong, 43));
    final frames = drainByteByByte(b.takeBytes());
    expect(frames.length, 5);
    expect(frames[0].type, frameHello);
    expect(frames[0].linkIdx, 7);
    expect(frames[0].connId, connId);
    expect(frames[1].type, frameData);
    expect(frames[1].offset, 1234567890123);
    expect(frames[1].payload, payload);
    expect(frames[2].offset, 999);
    expect(frames[3].nonce, 42);
    expect(frames[4].nonce, 43);
  });

  test('detects a corrupt DATA payload', () {
    final wire = encodeData(0, Uint8List.fromList([1, 2, 3, 4]));
    wire[wire.length - 1] ^= 0xff;
    final r = FrameReader()..push(wire);
    expect(r.next, throwsA(isA<StateError>()));
  });

  test('computes IEEE CRC32 (pinned to the peers)', () {
    final wire = encodeData(0, utf8.encode('123456789'));
    final crc = ByteData.view(wire.buffer).getUint32(11);
    expect(crc, 0xcbf43926);
  });

  test('L5 header round-trips and stops at the boundary', () {
    for (final h in [
      Header(kindNormal),
      Header(kindTunnel, target: 'api.anthropic.com:443'),
      Header(kindTunnel,
          target: '10.0.0.1:8080', ticket: Uint8List.fromList([9, 8, 7])),
    ]) {
      final b = BytesBuilder()
        ..add(encodeHeader(h))
        ..addByte(0x58);
      final out = readHeader(b.takeBytes())!;
      expect(out.header.kind, h.kind);
      expect(out.header.target, h.target);
    }
  });
}

File: ts/test/codecs.test.ts
import { describe, expect, it } from 'vitest';
import {
  FrameReader,
  FrameType,
  encodeAck,
  encodeData,
  encodeHello,
  encodeNonce,
} from '../src/frames.js';
import { encodeHeader, readHeader } from '../src/header.js';

// Feeds bytes to a FrameReader one byte at a time — the worst case a WebSocket can hand it — and
// collects every frame that falls out, proving reassembly across arbitrary chunk boundaries.
function drainByteByByte(bytes: Uint8Array) {
  const r = new FrameReader();
  const frames = [];
  for (const b of bytes) {
    r.push(Uint8Array.of(b));
    let f;
    while ((f = r.next()) !== null) frames.push(f);
  }
  return frames;
}

describe('redundant frames', () => {
  it('round-trips every frame type, reassembled byte by byte', () => {
    const connID = new Uint8Array(16).map((_, i) => i + 1);
    const payload = new Uint8Array(300).map((_, i) => i & 0xff); // spans the u16 length path
    const wire = new Uint8Array([
      ...encodeHello(connID, 7),
      ...encodeData(1234567890123n, payload),
      ...encodeAck(999n),
      ...encodeNonce(FrameType.Ping, 42n),
      ...encodeNonce(FrameType.Pong, 43n),
    ]);
    const frames = drainByteByByte(wire);
    expect(frames).toHaveLength(5);
    expect(frames[0]).toMatchObject({ type: FrameType.Hello, linkIdx: 7 });
    expect((frames[0] as { connID: Uint8Array }).connID).toEqual(connID);
    expect(frames[1]).toMatchObject({ type: FrameType.Data, offset: 1234567890123n });
    expect((frames[1] as { payload: Uint8Array }).payload).toEqual(payload);
    expect(frames[2]).toMatchObject({ type: FrameType.Ack, cumulative: 999n });
    expect(frames[3]).toMatchObject({ type: FrameType.Ping, nonce: 42n });
    expect(frames[4]).toMatchObject({ type: FrameType.Pong, nonce: 43n });
  });

  it('detects a corrupt DATA payload', () => {
    const wire = encodeData(0n, Uint8Array.of(1, 2, 3, 4));
    wire[wire.length - 1] ^= 0xff; // flip a payload byte; CRC no longer matches
    const r = new FrameReader();
    r.push(wire);
    expect(() => r.next()).toThrow(/corrupt/);
  });

  it('computes IEEE CRC32 (pinned to the Go/JVM peers)', () => {
    // crc32("123456789") === 0xCBF43926, the canonical IEEE check value.
    const data = new TextEncoder().encode('123456789');
    const wire = encodeData(0n, data);
    const crc = new DataView(wire.buffer).getUint32(11);
    expect(crc >>> 0).toBe(0xcbf43926);
  });
});

describe('L5 header', () => {
  it('round-trips service and ticket, stopping at the boundary', () => {
    for (const h of [
      { service: 'anthropic' },
      { service: 'greeter', ticket: Uint8Array.of(9, 8, 7) },
    ]) {
      const wire = new Uint8Array([...encodeHeader(h), 0x58]); // a payload byte follows
      const out = readHeader(wire);
      expect(out).not.toBeNull();
      expect(out!.header.service).toBe(h.service);
      expect(wire[out!.consumed]).toBe(0x58); // consumed exactly the header
    }
  });

  it('asks for more bytes when the header is incomplete', () => {
    const wire = encodeHeader({ service: 'echo' });
    expect(readHeader(wire.subarray(0, wire.length - 1))).toBeNull();
  });
});
